src/main/java/org/sagebionetworks/web/client/widget/entity/EntityBadge.java

```diff
@@ -131,12 +133,52 @@ public void setEntityBundle(EntityBundle eb) {
 			view.showPrivateIcon();
 		}
 
+		String annotationsHtml = null;
 		if (annotations != null) {
 			Map<String, AnnotationsValue> annotationsMap = annotations.getAnnotations();
 			if (!annotationsMap.isEmpty()) {
-				view.setAnnotations(getAnnotationsHTML(annotationsMap));
+				annotationsHtml = getAnnotationsHTML(annotationsMap);
 			}
 		}
+
+		// In experimental mode, check if there's a bound JSON Schema + check validity
+		if (DisplayUtils.isInTestWebsite(cookies)) {
+			final String finalAnnotationsHtml = annotationsHtml;
+			jsClient.getSchemaBinding(eb.getEntity().getId(), new AsyncCallback<JsonSchemaObjectBinding>() {
+				@Override
+				public void onSuccess(JsonSchemaObjectBinding result) {
+					boolean hasSchema = true;
+
+					jsClient.getSchemaValidationResultsWithMatchingEtag(eb.getEntity().getId(), eb.getEntity().getEtag(), new AsyncCallback<ValidationResults>() {
```

Comment:
Based on how this works, it seems unlikely that the etag we pass here is out of date, so I think this is fine.  If this was more likely (if the schema validation happened on hover, for example), then this could get caught in an infinite loop (since the call keeps retrying until the etags match).

Reply:
Good catch, I didn't think of a race condition occurring in this order. I think here it’s not worth the effort to guard against it, but if I have to reimplement it in SRC, I’ll be cognizant of it.